rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,217 @@
+always correlate outputs with inputs on subsequent resources, and also mitigates behavior of bad actors in the cluster
+as described
+in [Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556).
+
+For cases where the output does not contain necessary information to match against inputs, one may instead match against
+the spec of the resource:
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterImageTemplate
+metadata:
+  name: example-image-building-thing
+spec:
+  template:
+    apiVersion: my-builder/v1
+    kind: ImageBuilderThatDoesntExposeInputDetailsInItsOutput
+    metadata:
+      name: ...
+    spec:
+      sourceUrl: "$(sources.source.url)$"
+      sourceRevision: "$(sources.source.url)$"
+observedMatches:
+  - input: "$(sources.source.url)"              #evaluated against inputs in template context
+    output: "$(spec.sourceUrl)"                 #evaluated against resource spec
+  - input: "$(sources.source.revision)"         #evaluated against inputs in template context
+    output: "$(spec.sourceRevision)"            #evaluated against resource spec
+```
+
+Since the spec of the resource may change independently of the output, when the `outputMatches` makes any assertion
+against the `spec`, Cartographer must also ensure that the current output is a result of processing the current spec by
+making additional assertions:
+
+* Resource is healthy, as per [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+* Resource `observedGeneration == generation`
+
+If these assertions cannot be made, for example, if the resource does not implement the `observedGeneration` pattern,
+then Cartographer stops processing the resource and flags an error on it. Great care must also be taken that `healthy`
+status rules specified by the template author corresponds to the situation where the output is indeed a result of the
+spec with the current `observedGeneration`, otherwise outputs may not truly be the result of an input.
+
+Additionally, relying on the spec to make assertions imposes a further restriction to avoid stalling artifact
+propagation: Cartographer must not stamp out any further changes to the resource until it has finished processing the
+current spec completely. That is, it must have reached either a Healthy or Unhealthy status (as per
+[Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)).
+
+# Migration
+
+[migration]: #migration
+
+All templates must have `outputMatches` added to them.
+
+Any template which requires `outputMatches` against the resource spec must also be configured to identify success,
+failure and unknown states - which map to Healthy, Unhealthy and Unknown if the mechanism to be relied upon is
+[Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738).
+
+# Drawbacks
```

Comment:
Supply chains often rely on resources that continually report the latest good output (e.g. kpack). When such an object is processing new input, Cartographer is still able to read the output and use it to assert the proper shape of the following object.

WIth this proposal, Cartographer will need to adopt a caching strategy after reading good outputs of objects. Consider the case of a supply chain where resource B relies on resource A. When resource A has a good output, it will at some point get an update to its spec. It will begin process the new spec and `Healthy` will become unknown. As a result, Cartographer will no longer read resource A's output. That means that Cartographer will now be unable to assure that resource B has the proper spec, so resource B will be unreadable as well.

Reply:
I see. A cache, that is, or something like `lastGoodOtput` on the status in your [original RFC](https://github.com/vmware-tanzu/cartographer/pull/556)?

Reply:
Where is the cache stored? Assume that the controller pod will crash or be rescheduled at an inconvenient time; a new controller instance needs to be able to pickup and continue processing as if nothing happened.